### ISSUE TITLE
gRPC Support

### ISSUE THREAD

**Issue by ghostdogpr** [getkyo/kyo#390]
Support for protobuf and gRPC would drive adoption (at least for me).

It's definitely not easy as it involves code generation but maybe https://scalapb.github.io/docs/writing-plugins could be used like it is for zio-grpc and fs2-grpc.

There is currently the option of using zio-grpc with the zio interop but its performance is not great to start with (https://github.com/LesnyRumcajs/grpc_bench/discussions/441) so there might be an opportunity to shine.

**Comment by He-Pin** [comment]
Yes, looking forward to it, grpc_bench is a great place to show .

**Comment by steinybot** [comment]
This is still a WIP. I made some good progress on the code gen side of things. I haven't had much time in the last 2 weeks. I hope to get back to this in a week or so.

**Comment by steinybot** [comment]
Not sure what I changed but after implementing it as a library it is now significantly slower:
```
-----------------------------------------------------------------------------------------------------------------------------------------
| name                        |   req/s |   avg. latency |        90 % in |        95 % in |        99 % in | avg. cpu |   avg. memory |
-----------------------------------------------------------------------------------------------------------------------------------------
| scala_zio                   |   31949 |       29.93 ms |       66.57 ms |       76.40 ms |      132.59 ms |   100.5% |    315.89 MiB |
| scala_kyo                   |    7056 |      140.20 ms |      214.61 ms |      502.63 ms |         1.39 s |  101.77% |    333.34 MiB |
-----------------------------------------------------------------------------------------------------------------------------------------
```

**Comment by DamianReeves** [comment]
Any updates on this front?

**Comment by steinybot** [comment]
I haven't managed to dig into the performance issues. I created some benchmarks but that is as far as I got. I'm in the middle of merging the new core changes in.

Unfortunately I'm unexpectedly out of a job so I actually have less free time while I apply for jobs and do interviews etc. Happy to share what I have done and split the bounty if you want to help out.

**Comment by steinybot** [comment]
I'm back onto this now. I hope to have a draft PR up in a day or two.

**Comment by fwbrasil** [comment]
/bounty 500

Updating since this is much more work than I expected

**Comment by debaa98** [comment]
hey @fwbrasil is it like 700 or 500 ?

**Comment by fwbrasil** [comment]
It's 500 from Kyo's bounty + 250 from Kaizen. I believe @steinybot is working on this, though. We have other bounties open if you want to have a look: https://github.com/getkyo/kyo/issues?q=is%3Aissue%20state%3Aopen%20label%3A%22%F0%9F%92%8E%20Bounty%22

**Comment by steinybot** [comment]
I'm still working through the migration errors after updating from main. Hopefully not too much longer.

**Comment by steinybot** [comment]
Finally got it compiling again with the latest changes. I'll have to go back through it all again and remind myself where I was up to. I'll probably reach out for some help on the benchmark side of things in a week or so.

**Comment by steinybot** [comment]
Added cats to the benchmarks using https://github.com/typelevel/fs2-grpc:
```
[info] Benchmark                       Mode  Cnt      Score     Error  Units
[info] GrpcServerUnaryBench.forkCats  thrpt    5    524.112 ± 144.156  ops/s
[info] GrpcServerUnaryBench.forkKyo   thrpt    5    362.066 ±  30.976  ops/s
[info] GrpcServerUnaryBench.forkZIO   thrpt    5  10020.036 ± 556.540  ops/s
```
Seems like something might be wrong with the how I have setup the ZIO benchmarks unless it really is that fast.

**Comment by steinybot** [comment]
Lol yeah I was doing it wrong. I had the server setup in the ZLayer. Here are more comparable times:
```
[info] Benchmark                       Mode  Cnt    Score     Error  Units
[info] GrpcServerUnaryBench.forkCats  thrpt    5  460.361 ± 130.370  ops/s
[info] GrpcServerUnaryBench.forkKyo   thrpt    5  370.144 ±  51.291  ops/s
[info] GrpcServerUnaryBench.forkZIO   thrpt    5  394.379 ±  69.561  ops/s
```

**Comment by steinybot** [comment]
I haven't forgotten about this. I've been chipping away at it very slowly. Got quite a few tests now but there are some that have me a bit stumped which fail when run all at once which suggest a concurrency problem which is escaping me.

**Comment by steinybot** [comment]
Forgot I had already fixed that problem. I was using a SingleProducerSingleConsumer with multiple producers/consumers in my test 🙃.

Gettting somewhere now although this times out in my test:
```
        KyoApp.Unsafe.runAndBlock(Duration.Infinity)(put).getOrThrow
```
Why? It never seems to run the `put` effect.

**Comment by steinybot** [comment]
Finally found the root cause of a race condition (#1193) that had been blocking me for a while.

I've implemented pretty much everything although there are still quite a few little TODOs, tests, and benchmarks.

I just rewrote the benchmarks for gRPC to start the server once during setup and Kyo seems to be the fastest now although it is early days. Cats seems to be all over the show.

**Comment by steinybot** [comment]
```
sbt:kyoJVM> kyo-bench/jmh:run -wi 20 -i 10 -r 1 -w 1 -f 1 -t 1 -foe true -rf json GrpcServerUnary
...
[info] Benchmark                        Mode  Cnt     Score     Error  Units
[info] GrpcServerUnaryBench.catsBench  thrpt   10  8497.022 ± 563.772  ops/s
[info] GrpcServerUnaryBench.kyoBench   thrpt   10  9221.407 ± 213.381  ops/s
[info] GrpcServerUnaryBench.zioBench   thrpt   10  7822.059 ± 666.360  ops/s
```

**Comment by steinybot** [comment]
I'm hitting a resource leak in Netty now (https://github.com/grpc/grpc-java/issues/12082). It is in all 3 effect systems so not specific to this implementation. I don't think I'm doing anything weird in the benchmarks except not consuming the result. I wondered if it was a scalapb grpc issue but it's only a very small shim in this case.

**Comment by fwbrasil** [comment]
cool!!! gprc is definitely a significant expansion of Kyo's ecosystem! 🚀🚀

**Comment by He-Pin** [comment]
How about later submit to https://github.com/LesnyRumcajs/grpc_bench

**Comment by steinybot** [comment]
Yeah sorry for the delay, I only get burts of time to work on it and it has been a long process keeping up with the API changes and uncovering some issue. I'm trying to get the ergonomics of the metadata handling right and then it should be done.

**Comment by Sambit003** [comment]
> Yeah sorry for the delay, I only get burts of time to work on it and it has been a long process keeping up with the API changes and uncovering some issue. I'm trying to get the ergonomics of the metadata handling right and then it should be done.

Can I collaborate with you? As you've done pretty much works, so I will not open another PR to spam the repo. If you're okay with my proposal, I will start working on. Do let me know your decision.

**Comment by Rik-System-Root** [comment]
/attempt #390.
https://github.com/sindresorhus/ky/pull/821.

**Comment by y9206345-cmyk** [comment]
attempt #390

**Comment by merchantmoh-debug** [comment]
@EnzoGaillardSystems @wh0amibjm @y9206345-cmyk @Rik-System-Root 

What's the point? They don't check their PR requests. Mine has been up for over a month. 

This is a something I've noticed on this "bounties" on Algora. These companies put up these bounties - people like us do the work. And the PR requests sit on reviewed for weeks or months at a time if ever.

The same thing occurred with Golem - I made an MCP server and guess what ? so did 20 other people. Only for them to "cancel" the bounty.

The utter disrespect for other people's time & effort is frankly unprofessional. There needs to a better system. You put up a bounty - we make the code - you review it and we get paid ASAP. We're not beggars. We are doing work YOU are either too lazy to or just can't figure out on your own.

I'm busy writing a revolutionary programming language (Ark-Lang) and contributing to critical libraries (Iterators, OTEL, 3x perf Google Go - Zombie state fix Microsoft Autogen)

If I spend my time and effort to do your dirty work I expect to get a review and get money in my account. Not to be ignored. And what's more I can see that @fwbrasil is active. So there is no excuse.

So with all due respect to this team - give your heads a shake. I'm tired of seeing how you guys treat bounty hunters. The fact that you expect the work to be done and you DECIDE if you feel like reviewing it whenever (if ever) gives a clear imbalance of authority.

Basically; you reveal how humans treat each other without rules and laws to force us to be decent and considerate. Am I writing a monologue in an issue on Github? 

Yes I am. 

Because someone has to say something. We got paid nothing in prison - got treated like slaves. Paid slave wages.

But at least when I was in prison - I GOT PAID.

Smh.

**Comment by ayewo** [comment]
> [@EnzoGaillardSystems](https://github.com/EnzoGaillardSystems) [@wh0amibjm](https://github.com/wh0amibjm) [@y9206345-cmyk](https://github.com/y9206345-cmyk) [@Rik-System-Root](https://github.com/Rik-System-Root)
> 
> What's the point? They don't check their PR requests. Mine has been up for over a month.
> 
> This is a something I've noticed on this "bounties" on Algora. These companies put up these bounties - people like us do the work. And the PR requests sit unreviewed for weeks or months at a time if ever.
> 
> The same thing occurred with Golem - I made an MCP server and guess what ? so did 20 other people. Only for them to "cancel" the bounty.
> 
> The utter disrespect for other people's time & effort is frankly unprofessional. There needs to be a better system. You put up a bounty - we make the code - you review it and we get paid ASAP.
> 
> We're not beggars. We are doing work YOU are either too lazy to or just can't figure out on your own.
> 
> I'm busy writing a revolutionary programming language (Ark-Lang) and contributing to critical libraries (Iterators, OTEL, 3x perf Google Go - Zombie state fix Microsoft Autogen)
> 
> If I spend my time and effort to do your dirty work I expect to get a review and get money in my account.
> 
> Not to be ignored. And what's more I can see that [@fwbrasil](https://github.com/fwbrasil) is active. So there is no excuse.
> 
> So with all due respect to this team - give your heads a shake. I'm tired of seeing how you guys treat bounty hunters. The fact that you expect the work to be done and you DECIDE if you feel like reviewing it whenever (if ever) gives a clear imbalance of authority.
> 
> Basically; you reveal how humans treat each other without rules and laws to force us to be decent and considerate. Am I writing a monologue in an issue on Github?
> 
> Yes I am.
> 
> Because someone has to say something. We got paid nothing in prison - got treated like slaves. Paid slave wages.
> 
> But at least when I was in prison - I GOT PAID.
> 
> Smh.


Sorry you feel that way @merchantmoh-debug but the simple truth is that your PR is several months late.

Here’s the Discord [announcement](https://discord.com/channels/1087005439859904574/1087005440623259681/1350275876029333545) by @fwbrasil 11 months ago pausing the bounty program because it didn’t attract enough long term contributors. It effectively withdrew all unclaimed bounties.

The Algora website doesn’t do a good job of hiding/removing stale bounties which is probably why you and others are still able to land here.

In future, it’s important to first establish communication with the maintainers _before_ opening a PR.

**Comment by justinleeyang** [comment]
I can take bounty #390. I will share a scoped implementation plan and first PR shortly.

**Comment by justinleeyang** [comment]
Quick follow-up on #390: I am still actively working on this. Please let me know if you want me to narrow scope to a smaller first PR.

**Comment by ryan-the-zilla** [comment]
@ghostdogpr Claiming gRPC Support bounty (00)!\n\nI'll implement full gRPC support with:\n- Code-first and proto-first approaches\n- Streaming support\n- Integration with Kyo effect system\n\nETA: 24-48h\n\nWallet: AqE264DnKyJci9kV4t3eYhDtFB3H88HQusWtH5odSqHM\n\nStarting now. 🚀

**Comment by schullegerhard** [comment]
It looks like there’s already been solid exploration here (especially around naive server wiring and early codegen work), but a few important gaps and potential issues stand out from the discussion and benchmarks:

**Comment by schullegerhard** [comment]
1. Per-request runtime execution is likely the main bottleneck

Right now each request does:

```
val fiber = Grpcs.run(Grpcs.init(serviceImpl.sayHello(request)))
IOs.run(fiber.onComplete { ... })

```

Suggested fix: reuse a single runtime

Instead of running the runtime inside the handler, initialize it once and reuse it:
```
object GrpcRuntime:
  val runtime = IOs.init() // or equivalent long-lived runtime

  def unsafeRun[T](v: T < IOs): Unit =
    IOs.run(v) // but avoid nesting this per request if possible
```

Then in your handler:
```
ServerCalls.asyncUnaryCall((request, observer) => {
  val effect = serviceImpl.sayHello(request)

  GrpcRuntime.unsafeRun {
    Grpcs.init(effect).map { fiber =>
      fiber.onComplete { result =>
        IOs.attempt(result).map(scalapb.grpc.Grpc.completeObserver(observer))
      }
    }
  }
})
```

Even better: avoid Fiber entirely if unary:
```
ServerCalls.asyncUnaryCall((request, observer) => {
  GrpcRuntime.unsafeRun {
    Aborts.run(serviceImpl.sayHello(request)).map {
      case Right(reply) =>
        observer.onNext(reply)
        observer.onCompleted()
      case Left(err) =>
        observer.onError(err)
    }
  }
})
```

2. Avoid Try conversions and double-wrapping

This part is costly:
```
.map(_.toTry)
.map(_.transform(_.fold(Fiber.fail, Fiber.value)))
```

Suggested simplification
```
def init[T: Flat](v: => T < Grpcs): Fiber[T] < IOs =
  Fibers.init {
    Aborts.run(v).map {
      case Right(value) => value
      case Left(err)    => throw err
    }
  }
```
Or better, don’t convert at all unless required by gRPC boundary.

3. Callback → effect bridge can be simplified

Current:
```
fiber.onComplete { reply =>
  IOs.attempt(reply).map(scalapb.grpc.Grpc.completeObserver(observer))
}
```

Suggested version
```
fiber.onComplete {
  case Success(reply) =>
    observer.onNext(reply)
    observer.onCompleted()
  case Failure(err) =>
    observer.onError(err)
}
```
Or fully inside effect:
```
Aborts.run(effect).map {
  case Right(reply) =>
    observer.onNext(reply)
    observer.onCompleted()
  case Left(err) =>
    observer.onError(err)
}
```

4. Executor choice is likely hurting performance

This:
```
builder.executor(Executors.newWorkStealingPool(i_threads))
```

Suggested experiments

Try:
```
builder.directExecutor()
```
or:
```
builder.executor(Executors.newFixedThreadPool(i_threads))
```

**Comment by CuboYe** [comment]
/claim 390

Claiming territory for a scoped fix attempt now. If maintainer preference differs, I can adapt quickly.

**Comment by TeapoyY** [comment]
I'll work on this for the  bounty.

I'll implement gRPC support for Kyo following the approach mentioned in the issue - using scalapb for code generation similar to zio-grpc and fs2-grpc.

@fwbrasil @calvinlfer @algora-pbc - claiming this bounty.

**Comment by themachinecorp** [comment]
/attempt #390

I would like to work on this gRPC Support feature for Kyo. My plan is to:
1. Explore the existing Kyo codebase to understand the architecture
2. Research scalapb (ScalaPB) plugin system as referenced in the issue
3. Implement basic protobuf code generation support for Kyo
4. Create a gRPC server/client example demonstrating native integration
5. Submit a PR with the implementation

Looking forward to implementing this feature!

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by alihani714** [comment]
/attempt #390

### Implementation Plan
1. Research the codebase to understand existing patterns.
2. Implement the missing feature/fix with high code quality.
3. Ensure comprehensive test coverage.

*Available to start now and deliver quickly.*

**Comment by shehryarakif** [comment]
// For the benchmark module
libraryDependencies ++= Seq(
  "dev.zio"             %% "zio-schema"               % "1.0.0",
  "dev.zio"             %% "zio-schema-protobuf"      % "1.0.0",
  "com.thesamet.scalapb" %% "scalapb-runtime"         % "0.11.17",
  "org.openjdk.jmh"      % "jmh-core"                 % "1.37",
  "org.openjdk.jmh"      % "jmh-generator-annprocess" % "1.37",
  "ch.qos.logback"       % "logback-classic"          % "1.5.6" % Optional
)

syntax = "proto3";

package benchmark;

message Person {
  string name = 1;
  int32 age = 2;
  string email = 3;
}

message AddressBook {
  repeated Person people = 1;
  int64 timestamp = 2;
}

PB.targets in Compile := Seq(
  scalapb.gen() -> (sourceManaged in Compile).value / "scalapb"

// src/main/scala/benchmark/ZioSchemas.scala
package benchmark

import zio.schema._
import zio.schema.annotation._

case class Person(name: String, age: Int, email: String)
object Person {
  implicit val schema: Schema[Person] = DeriveSchema.gen[Person]
}

case class AddressBook(people: Chunk[Person], timestamp: Long)
object AddressBook {
  implicit val schema: Schema[AddressBook] = DeriveSchema.gen[AddressBook]
}

// src/main/scala/benchmark/ProtobufBenchmark.scala
package benchmark

import org.openjdk.jmh.annotations._
import java.util.concurrent.TimeUnit
import scalapb.GeneratedMessage
import zio.schema.codec.ProtobufCodec
import zio.Chunk

@State(Scope.Thread)
@BenchmarkMode(Array(Mode.Throughput, Mode.AverageTime))
@OutputTimeUnit(TimeUnit.MICROSECONDS)
@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
@Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
@Fork(1)
class ProtobufBenchmark {

  // ---------- Test data ----------
  private val person: Person = Person("Alice Wonder", 42, "alice@example.com")
  private val addressBook: AddressBook = AddressBook(
    Chunk.fill(100)(person),
    System.currentTimeMillis()
  )

  // ---------- ScalaPB encoded bytes ----------
  private lazy val scalapbPersonBytes: Array[Byte] = ???
  private lazy val scalapbAddressBookBytes: Array[Byte] = ???

  // ---------- ZIO Schema encoded bytes ----------
  private lazy val zioPersonBytes: Array[Byte] = {
    val codec = ProtobufCodec(Person.schema)
    codec.encode(person).toArray
  }
  private lazy val zioAddressBookBytes: Array[Byte] = {
    val codec = ProtobufCodec(AddressBook.schema)
    codec.encode(addressBook).toArray
  }

  // ---------- ScalaPB benchmarks ----------
  @Benchmark
  def scalapb_serialize_person(): Array[Byte] =
    person.toByteArray

  @Benchmark
  def scalapb_deserialize_person(): Person =
    Person.parseFrom(scalapbPersonBytes)

  @Benchmark
  def scalapb_serialize_addressbook(): Array[Byte] =
    addressBook.toByteArray

  @Benchmark
  def scalapb_deserialize_addressbook(): AddressBook =
    AddressBook.parseFrom(scalapbAddressBookBytes)

  // ---------- ZIO Schema benchmarks ----------
  @Benchmark
  def zio_serialize_person(): Array[Byte] = {
    val codec = ProtobufCodec(Person.schema)
    codec.encode(person).toArray
  }

  @Benchmark
  def zio_deserialize_person(): Person = {
    val codec = ProtobufCodec(Person.schema)
    codec.decode(Chunk.fromArray(zioPersonBytes)).toOption.get
  }

  @Benchmark
  def zio_serialize_addressbook(): Array[Byte] = {
    val codec = ProtobufCodec(AddressBook.schema)
    codec.encode(addressBook).toArray
  }

  @Benchmark
  def zio_deserialize_addressbook(): AddressBook = {
    val codec = ProtobufCodec(AddressBook.schema)
    codec.decode(Chunk.fromArray(zioAddressBookBytes)).toOption.get
  }
}

sbt "benchmark/jmh:run -i 3 -wi 3 -f 1 -t 1 .*ProtobufBenchmark.*"

sbt "benchmark/jmh:run -prof gc .*ProtobufBenchmark.*"

// project/plugins.sbt
addSbtPlugin("com.thesamet" % "sbt-protoc" % "1.0.6")

// build.sbt
PB.targets in Compile := Seq(
  scalapb.gen() -> (sourceManaged in Compile).value / "scalapb"
)

libraryDependencies ++= Seq(
  "com.thesamet.scalapb" %% "scalapb-runtime" % "0.11.17" % "protobuf",
  "dev.zio"             %% "zio-schema"       % "1.0.0",
  "dev.zio"             %% "zio-schema-protobuf" % "1.0.0"
)

// src/main/scala/benchmark/GenerateScalaPBBytes.scala
package benchmark

import java.nio.file.{Files, Paths}
import scala.pickling.`package` // not needed; just use java.io

object GenerateScalaPBBytes {
  def main(args: Array[String]): Unit = {
    // Same test data as in the benchmark
    val person = Person("Alice Wonder", 42, "alice@example.com")
    val addressBook = AddressBook(people = Seq.fill(100)(person), timestamp = System.currentTimeMillis())

    // Serialize with ScalaPB
    val personBytes = person.toByteArray
    val addressBookBytes = addressBook.toByteArray

    // Write to files (or print as Scala code)
    val personArrayLiteral = personBytes.map(b => f"$b.toByte").mkString("Array(", ", ", ")")
    val addressBookArrayLiteral = addressBookBytes.map(b => f"$b.toByte").mkString("Array(", ", ", ")")

    val code =
      s"""|// Auto‑generated byte arrays for ScalaPB benchmark
          |object ScalaPBPrecomputed {
          |  val personBytes: Array[Byte] = $personArrayLiteral
          |  val addressBookBytes: Array[Byte] = $addressBookArrayLiteral
          |}
          |""".stripMargin

    Files.write(Paths.get("ScalaPBPrecomputed.scala"), code.getBytes("UTF-8"))
    println("Generated ScalaPBPrecomputed.scala")
  }
}

// src/main/scala/benchmark/ProtobufBenchmark.scala
package benchmark

import org.openjdk.jmh.annotations._
import java.util.concurrent.TimeUnit
import zio.schema.codec.ProtobufCodec
import zio.Chunk

@State(Scope.Thread)
@BenchmarkMode(Array(Mode.Throughput, Mode.AverageTime))
@OutputTimeUnit(TimeUnit.MICROSECONDS)
@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
@Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
@Fork(1)
class ProtobufBenchmark {

  // Test data (same for both codecs)
  private val person: Person = Person("Alice Wonder", 42, "alice@example.com")
  private val addressBook: AddressBook = AddressBook(
    people = (1 to 100).map(_ => person).toList,
    timestamp = System.currentTimeMillis()
  )

  // Precomputed ScalaPB bytes (from the generated object)
  private val scalapbPersonBytes: Array[Byte] = ScalaPBPrecomputed.personBytes
  private val scalapbAddressBookBytes: Array[Byte] = ScalaPBPrecomputed.addressBookBytes

  // Precomputed ZIO Schema bytes (cached for deserialization benchmarks)
  private lazy val zioPersonBytes: Array[Byte] = {
    val codec = ProtobufCodec(Person.schema)
    codec.encode(person).toArray
  }
  private lazy val zioAddressBookBytes: Array[Byte] = {
    val codec = ProtobufCodec(AddressBook.schema)
    codec.encode(addressBook).toArray
  }

  // ---------- ScalaPB benchmarks ----------
  @Benchmark
  def scalapb_serialize_person(): Array[Byte] =
    person.toByteArray

  @Benchmark
  def scalapb_deserialize_person(): Person =
    Person.parseFrom(scalapbPersonBytes)

  @Benchmark
  def scalapb_serialize_addressbook(): Array[Byte] =
    addressBook.toByteArray

  @Benchmark
  def scalapb_deserialize_addressbook(): AddressBook =
    AddressBook.parseFrom(scalapbAddressBookBytes)

  // ---------- ZIO Schema benchmarks ----------
  @Benchmark
  def zio_serialize_person(): Array[Byte] = {
    val codec = ProtobufCodec(Person.schema)
    codec.encode(person).toArray
  }

  @Benchmark
  def zio_deserialize_person(): Person = {
    val codec = ProtobufCodec(Person.schema)
    codec.decode(Chunk.fromArray(zioPersonBytes)).toOption.get
  }

  @Benchmark
  def zio_serialize_addressbook(): Array[Byte] = {
    val codec = ProtobufCodec(AddressBook.schema)
    codec.encode(addressBook).toArray
  }

  @Benchmark
  def zio_deserialize_addressbook(): AddressBook = {
    val codec = ProtobufCodec(AddressBook.schema)
    codec.decode(Chunk.fromArray(zioAddressBookBytes)).toOption.get
  }
}

assert(java.util.Arrays.equals(zioPersonBytes, scalapbPersonBytes))

sbt "benchmark/jmh:run -i 3 -wi 3 -f 1 -t 1 .*ProtobufBenchmark.*"
)

**Comment by rrafike970-star** [comment]
/attempt #390

Subject: Verified Solution for Kyo-gRPC Bridge (Fibers & Aborts support)

Hi @kitlangton,

I have implemented a functional bridge between Kyo and gRPC StreamObserver. Unlike previous attempts, this solution is already tested and verified locally.

Implementation Details:
Fiber Integration: Uses Fibers.init for non-blocking execution, detaching from the gRPC thread.

Error Handling: Correctly maps Aborts[Throwable] to observer.onError and handles the result lifecycle.

Scala 3 Ready: Uses Flat[T] evidence to ensure type safety within the Kyo effect system.

Version Resilience: Uses robust pattern matching to handle both Result and Either types.

Local Verification Results:
Plaintext
Starting Kyo-gRPC Bridge Test...
-> Received: Success! Data Compressed.
-> Connection Closed Successfully.
[success] Total time: 6 s
The Solution:
Scala
import kyo.*
import io.grpc.stub.StreamObserver

object KyoBridge:
  def bridge[T](effect: T < (Fibers & Aborts[Throwable]), observer: StreamObserver[T])(using Flat[T]): Unit =
    val handledEffect = Aborts[Throwable].run(effect)
    val fiberEffect = Fibers.init(handledEffect)
    
    IOs.run(fiberEffect.map(_.onComplete { 
      case Right(outcome) =>
        outcome match
          case Right(value) => 
            observer.onNext(value.asInstanceOf[T])
            observer.onCompleted()
          case Left(ex: Throwable) => 
            observer.onError(ex)
          case value if value != null => 
            observer.onNext(value.asInstanceOf[T])
            observer.onCompleted()
          case _ => 
            observer.onError(new RuntimeException("Empty outcome"))
      case Left(fiberEx: Throwable) =>
        observer.onError(fiberEx)
      case _ =>
        observer.onError(new RuntimeException("Fiber failed"))
    }))
Best regards,
Mohamed Rafik

**Comment by whutlichao** [comment]
## 🎯 Bounty Application

**Developer**: RoyceLee
**Bounty**: $500
**Stack**: TypeScript/Java + gRPC + API Integration

---

### 👤 Relevant Experience

Full-stack developer with extensive experience in API integration and distributed systems.

**Related Projects**:
- ✅ Built gRPC-based microservices architecture
- ✅ Implemented service mesh with Istio
- ✅ Integrated REST/gRPC APIs in TypeScript/Java
- ✅ Developed streaming RPC interfaces

---

### 📋 Understanding

This task requires implementing gRPC support for the Kyo framework:
1. Add gRPC client/server capabilities
2. Support streaming RPCs
3. Integrate with existing Kyo effects system
4. Provide type-safe gRPC wrappers

---

### 💡 Technical Approach

**Implementation Plan**:
1. Analyze Kyo's effect system (2h)
2. Design gRPC integration layer (2h)
3. Implement gRPC client stubs (3h)
4. Implement gRPC server handlers (3h)
5. Add streaming support (2h)
6. Write comprehensive tests (2h)
7. Documentation (1h)

**Total**: 15 hours | **Delivery**: 3-5 days

---

### ✅ Deliverables

- [ ] gRPC client implementation
- [ ] gRPC server implementation
- [ ] Streaming RPC support
- [ ] Test coverage >80%
- [ ] API documentation
- [ ] Usage examples

---

### ⚡ Availability

**Start**: Immediately
**Communication**: GitHub Issue / Discord
**Payment**: Algora ✓

**Ready to start now!** 🚀

---

Looking forward to contributing to Kyo!

**Comment by whutlichao** [comment]
## 🎯 Bounty Application

**Developer**: whutlichao
**Bounty**: $500
**Stack**: TypeScript/Java + gRPC + API Integration

---

### 👤 Relevant Experience

Full-stack developer with extensive experience in API integration and distributed systems.

**Related Projects**:
- ✅ Built gRPC-based microservices architecture
- ✅ Implemented service mesh with Istio
- ✅ Integrated REST/gRPC APIs in TypeScript/Java
- ✅ Developed streaming RPC interfaces

---

### 📋 Understanding

This task requires implementing gRPC support for the Kyo framework:
1. Add gRPC client/server capabilities
2. Support streaming RPCs
3. Integrate with existing Kyo effects system
4. Provide type-safe gRPC wrappers

---

### 💡 Technical Approach

**Implementation Plan**:
1. Analyze Kyo's effect system (2h)
2. Design gRPC integration layer (2h)
3. Implement gRPC client stubs (3h)
4. Implement gRPC server handlers (3h)
5. Add streaming support (2h)
6. Write comprehensive tests (2h)
7. Documentation (1h)

**Total**: 15 hours | **Delivery**: 3-5 days

---

### ✅ Deliverables

- [ ] gRPC client implementation
- [ ] gRPC server implementation
- [ ] Streaming RPC support
- [ ] Test coverage >80%
- [ ] API documentation
- [ ] Usage examples

---

### ⚡ Availability

**Start**: Immediately
**Communication**: GitHub Issue / Discord
**Payment**: Algora ✓

**Ready to start now!** 🚀

---

Looking forward to contributing to Kyo!

**Comment by whutlichao** [comment]
## Bounty Application

Hello Kyo maintainers — I'd like to implement gRPC support for Kyo.

**GitHub account:** `whutlichao`  
**Relevant stack:** Java, gRPC, Netty, TypeScript, Python

### Why I'm a good fit
I have hands-on experience with gRPC in Java/Netty environments and understand the protobuf ecosystem. I'm familiar with the performance considerations mentioned in the issue (referencing grpc_bench).

### My understanding
This issue requests native gRPC support for Kyo, similar to zio-grpc but with better performance. Key requirements:
- Protobuf code generation integration
- gRPC client/server implementation
- Integration with Kyo's fiber/effect system
- Performance competitive with or better than zio-grpc

### Proposed execution plan
1. **Study existing solutions** - Analyze zio-grpc, fs2-grpc implementations
2. **Design** - Define Kyo-grpc module structure and API
3. **Implementation** - Build gRPC integration using Java gRPC + Kyo effects
4. **Benchmarks** - Compare against zio-grpc using grpc_bench
5. **Examples** - Create sample applications demonstrating usage
6. **Documentation** - Comprehensive docs and API reference

### Delivery expectations
- **Start**: Immediately
- **Estimated turnaround**: 10-14 days for initial release
- **Deliverables**: kyo-grpc module, benchmarks, examples, docs

Excited to bring gRPC support to Kyo with a focus on performance!

**Comment by robinchen0127-star** [comment]
Atlas here. I read through the thread and saw @steinybot mention they were open to sharing work / splitting the bounty. I can help on a scoped chunk instead of adding another generic `/attempt`.

Proposed contribution scope:
1. reproduce the current grpc_bench baseline against main
2. isolate where the current Kyo integration loses throughput / adds allocations
3. own the benchmark + profiling pass (cpu / alloc / latency)
4. if useful, submit a focused PR on perf fixes or integration cleanup rather than broad codegen churn

If that split sounds useful, I will start with the benchmark/profiling track and post concrete numbers before touching API shape.

**Comment by rishi-jat** [comment]
Hi @ghostdogpr  @fwbrasil ,

I’ve gone through the thread, prior attempts, and the benchmark data. The main gap here doesn’t look like “missing gRPC support” — it’s the integration quality between gRPC’s execution model and Kyo’s effect system, especially around fibers, scheduling, and allocation behavior.

Instead of proposing a full rewrite or another broad attempt, I’d like to take a scoped, engineering-first approach:

Proposed focus (initial phase):
- Reproduce current grpc_bench results against main
- Profile the Kyo implementation (CPU, allocations, latency)
- Identify where throughput is lost:
- Fiber lifecycle overhead (init/run boundaries)
- IOs.run usage inside hot paths
- gRPC thread ↔ Kyo scheduler interaction
- Backpressure / observer completion behavior
- Share concrete findings before changing APIs

Goal:
Establish a clear performance baseline and bottleneck map so any further work (codegen, API design, streaming support) is built on something that can actually compete with akka/pekko/fs2.

If useful, I’m happy to collaborate with @steinybot and others already working here — especially splitting along benchmarking/profiling vs implementation so we avoid duplicating effort.

If this direction aligns with what maintainers expect for this bounty, I’ll start with the benchmarking + profiling pass and post results.

Also, please confirm if:
- Matching or exceeding zio-grpc performance is an explicit acceptance requirement
- There are any constraints on implementation approach (e.g., required use of ScalaPB plugins vs custom codegen)

Thanks.

**Comment by themachinehf** [comment]
Hi @getkyo — I'd like to reopen this PR and address any issues. Can you let me know what needs to be changed?

**Comment by themachinehf** [comment]
/claim #390

Hi @getkyo — re-creating PR #1499 as a fresh PR with resolved merge conflicts. The kyo-grpc module adds fiber-native gRPC support. Ready for review.

**Comment by webhop123** [comment]
I pushed a refreshed branch here: https://github.com/webhop123/kyo/tree/codex/kyo-grpc-refresh

This brings the gRPC work back onto current main, resolves the merge conflicts, fixes Windows/codegen line-ending issues, and tightens streaming completion behavior around final status propagation/readiness waits.

Validated locally with:
- `sbt --no-server kyo-grpc-core/test kyo-grpc-code-gen/test kyo-grpc-e2e/test`

Keeping this as a branch for now since there are already open attempts/PRs around this issue, but it should be easy to compare or cherry-pick if useful.

**Comment by MilosM348** [comment]
Hi @fwbrasil — I'd like to attempt this. Before I write any code I want to make sure I'm not stepping on @steinybot's #1334 or @webhop123's #1526.

**Quick questions to scope my work:**

1. Is #1334 or #1526 being shepherded to merge? If either is the path forward, happy to hold off and just review/test instead.
2. If you'd rather have a fresh attempt, what's the preferred scope for a first PR?
   - **a)** All four RPC shapes (unary, server-stream, client-stream, bidi) in one module
   - **b)** Unary + server-stream first, streaming-bidi as a follow-up
   - **c)** Something else
3. For codegen, do you prefer:
   - **a)** A `scalapb` adapter that produces Kyo-style stubs (faster to ship)
   - **b)** A native sbt plugin for `kyo-grpc-codegen` reading `.proto` directly (cleaner, more work)
4. Should the error type be `Abort[io.grpc.Status]` (thin wrapper) or a Kyo-native sealed ADT (`KyoStatus`) that maps gRPC codes?

**My approach if I get the green light:** new `kyo-grpc` module with `Server`/`Client`/`Channel`/`Status`/`Metadata`, bridging gRPC's `StreamObserver` to Kyo's `Stream`. Suspends in `Async`, errors flow through `Abort[Status]`, deadlines via a `Local[Deadline]`. Reference points: `zio-grpc`, `fs2-grpc`. Target PR size ~2,000 LOC excluding generated code, split into a stack if larger.

ETA: ~10–14 days from green-light. Happy to share a design doc for sign-off before opening any PR.

/attempt #390

**Comment by webhop123** [comment]
> Hi [@fwbrasil](https://github.com/fwbrasil) — I'd like to attempt this. Before I write any code I want to make sure I'm not stepping on [@steinybot](https://github.com/steinybot)'s [#1334](https://github.com/getkyo/kyo/pull/1334) or [@webhop123](https://github.com/webhop123)'s [#1526](https://github.com/getkyo/kyo/pull/1526).
> 
> **Quick questions to scope my work:**
> 
> 1. Is [Kyo grpc #1334](https://github.com/getkyo/kyo/pull/1334) or [Refresh gRPC support against current main #1526](https://github.com/getkyo/kyo/pull/1526) being shepherded to merge? If either is the path forward, happy to hold off and just review/test instead.
> 2. If you'd rather have a fresh attempt, what's the preferred scope for a first PR?
>    
>    * **a)** All four RPC shapes (unary, server-stream, client-stream, bidi) in one module
>    * **b)** Unary + server-stream first, streaming-bidi as a follow-up
>    * **c)** Something else
> 3. For codegen, do you prefer:
>    
>    * **a)** A `scalapb` adapter that produces Kyo-style stubs (faster to ship)
>    * **b)** A native sbt plugin for `kyo-grpc-codegen` reading `.proto` directly (cleaner, more work)
> 4. Should the error type be `Abort[io.grpc.Status]` (thin wrapper) or a Kyo-native sealed ADT (`KyoStatus`) that maps gRPC codes?
> 
> **My approach if I get the green light:** new `kyo-grpc` module with `Server`/`Client`/`Channel`/`Status`/`Metadata`, bridging gRPC's `StreamObserver` to Kyo's `Stream`. Suspends in `Async`, errors flow through `Abort[Status]`, deadlines via a `Local[Deadline]`. Reference points: `zio-grpc`, `fs2-grpc`. Target PR size ~2,000 LOC excluding generated code, split into a stack if larger.
> 
> ETA: ~10–14 days from green-light. Happy to share a design doc for sign-off before opening any PR.
> 
> /attempt [#390](https://github.com/getkyo/kyo/issues/390)

#1526 is refreshed against current main and ready for review.

I think the most useful next step is review/testing on #1526 unless maintainers prefer a different direction.

**Comment by SAVAMforWIN** [comment]
Atlas here. I read the thread and I do not want to duplicate the existing large implementation tracks blindly. My plan is scoped and practical: 1) reproduce the current grpc_bench baseline against current main, 2) inspect the refreshed branches/PRs already mentioned in the thread, 3) isolate whether the remaining gap is codegen correctness, streaming completion semantics, or pure performance overhead, and 4) only then prepare either a focused fix PR or a reviewable benchmark/profiling contribution with clear payout-relevant value. /attempt #390

**Comment by arbazkhan971** [comment]
I implemented a focused follow-up for #390 in PR #1540: send first bidi-streaming request without waiting on readiness, then gate subsequent sends and abort safely on call close. In case you still accept additional attempts before claim PR #1334 is merged.

**Comment by zxzok** [comment]
I opened a narrow follow-up against this branch: https://github.com/webhop123/kyo/pull/2

It keeps the first-request fast path, but avoids a remaining readiness race by clearing the signal before checking `call.isReady` for subsequent requests. It also uses the ready/completion race before later sends, so early server close does not leave the sender waiting forever.

Validated locally on top of this PR with:

`JAVA_HOME=/opt/homebrew/opt/openjdk@21/libexec/openjdk.jdk/Contents/Home PATH=/opt/homebrew/opt/openjdk@21/bin:/opt/homebrew/bin:/usr/bin:/bin SBT_OPTS='-Xms3G -Xmx4G -Xss10M -XX:MaxMetaspaceSize=512M -XX:ReservedCodeCacheSize=128M -Dfile.encoding=UTF-8' sbt --no-server "kyo-grpc-e2e/testOnly kyo.grpc.ServiceTest"`

Result: 26 tests passed.

**Comment by zxzok** [comment]
/attempt #390

I am contributing a focused follow-up to the current gRPC refresh path rather than duplicating the full implementation.

What I submitted:
- Follow-up PR against the active refresh branch: https://github.com/webhop123/kyo/pull/2
- Upstream PR note: https://github.com/getkyo/kyo/pull/1526#issuecomment-4434962335

Scope:
- Fix a remaining streaming readiness race in `ClientCall`.
- Keep the first-request fast path so client-streaming and bidi calls do not wait for readiness before the first message.
- For subsequent messages, clear the readiness signal before checking `call.isReady`, then wait on either readiness or call completion before sending. This avoids losing an `onReady` callback and avoids hanging when the server closes early.

Validation:
- `kyo-grpc-e2e/testOnly kyo.grpc.ServiceTest`
- 26 tests passed locally on top of the active refresh branch.

**Comment by zxzok** [comment]
Additional validation for the follow-up PR:

- `kyo-grpc-core/test`: 207 tests passed
- `kyo-grpc-e2e/test`: 26 tests passed

This was run locally on the follow-up branch with Java 21 and `sbt --no-server`.

**Comment by AUsernameForYou** [comment]
I'd like to work on this bounty — adding gRPC support to Kyo.

**Approach:**
1. Evaluate scalapb code generation as suggested in the issue
2. Design a Kyo-native gRPC API that leverages Kyo's effect system (similar to how kyo-sttp/kyo-tapir work)
3. Implement server and client stubs generated from .proto files
4. Add streaming support (bidirectional streaming is key for gRPC)
5. Benchmark against zio-grpc to validate the performance advantage
6. Add tests and documentation

I have experience with gRPC in Scala (zio-grpc, fs2-grpc) and am excited to build a Kyo-idiomatic solution. Ready to start once scope is confirmed!

**Comment by akamabu** [comment]
/attempt #390

Adding `deadline` field to `RequestOptions` so per-call gRPC deadlines can be set via `CallOptions.withDeadline`. The deadline is now applied when creating the `ClientCall` for all four RPC patterns (unary, client streaming, server streaming, bidi streaming).

**Comment by akamabu** [comment]
Opened PR: https://github.com/getkyo/kyo/pull/1542

**Comment by demo112** [comment]
/attempt #390

I would like to work on the gRPC support bounty. I have experience with gRPC and Scala.

**Comment by weilixiong** [comment]
/attempt #390

Scoped follow-up on top of the active gRPC refresh work: I opened PR #1544 to address the remaining client-streaming / bidi-streaming readiness race in `ClientCall`.

Scope:
- send the first request without waiting on readiness
- for later sends, wait on either readiness or call completion
- avoid stranding the sender when the server closes early

PR: https://github.com/getkyo/kyo/pull/1544

**Comment by mt211211** [comment]
/attempt #390

Focused follow-up to the active gRPC implementation rather than a duplicate full implementation: https://github.com/frank-ih/kyo/pull/1

I found a bidi-streaming edge case where the client waits for response headers before sending the outbound stream. Servers that only flush initial metadata after receiving the first request can deadlock. The follow-up patch aligns bidi with the client-streaming path and adds an e2e regression for delayed response headers.

**Comment by mt211211** [comment]
/claim #390

Public claim linkage for the focused follow-up PR: https://github.com/frank-ih/kyo/pull/1

This is scoped to the bidi-streaming delayed-headers deadlock edge case on the active gRPC implementation branch.

**Comment by jafethalejandroruizrocha-source** [comment]
/attempt #1545 Hi! We noticed this issue has been inactive for a while. My development team (@alexeing7-tech, @LCabrerag) and I would love to take over and complete the gRPC support implementation. We have reviewed the previous notes and are ready to work on it.

**Comment by zhaog100** [comment]
👋 I'd like to work on this bounty. I have experience with gRPC and protobuf in Scala.

**Comment by john-iceverg** [comment]
I'd like to work on this bounty too. If it is not possible, can you  refer me to the other bounty please?

**Comment by LindaLiu2013** [comment]
Working now

**Comment by LindaLiu2013** [comment]
Protobuf HTTP bodies in kyo-http

**Comment by LindaLiu2013** [comment]
I opened a first-step PR toward gRPC support here: https://github.com/getkyo/kyo/pull/1548

It adds schema-backed Protobuf request/response body support in `kyo-http`, including route DSL support, encode/decode handling, OpenAPI media type reporting, and tests.

This is not a complete gRPC implementation yet, but it should provide a useful foundation for the transport/Protobuf part of #390.

**Comment by dimscm** [comment]
/attempt #390

Implementation plan:
1) Add protobuf codegen integration path for Kyo modules.
2) Introduce gRPC transport support with minimal API surface.
3) Add docs + sample + tests/bench notes.
4) Submit PR with /claim #390 in body.

**Comment by ycj52874267-debug** [comment]
Hi! I'd like to work on this bounty, but I want to clarify the expected scope before starting so I keep the PR reviewable and useful.

My proposed approach:

1. Start with a small design note comparing the intended Kyo gRPC API shape with existing ScalaPB plugin patterns, especially how zio-grpc / fs2-grpc structure generated service traits, clients, and streaming methods.
2. Implement a minimal ScalaPB codegen plugin prototype that can generate Kyo-friendly client/server bindings for a small unary service first.
3. Add a tiny example proto + generated integration test so the first PR proves the design without trying to cover every gRPC feature at once.
4. Follow up with streaming support and performance-focused work once the maintainer agrees with the API shape.

Before I start, could you confirm whether the first acceptable PR should target:

- A design/prototype PR with unary RPC support only, or
- Full unary + server streaming + client streaming + bidirectional streaming support in one PR?

Also, is there a preferred module name or package layout for this integration?

GitHub: @ycj52874267-debug

**Comment by circlemouth** [comment]
Hi! I'm looking into this issue and would like to confirm the intended scope before starting implementation.

Would a useful first PR be a JVM-only `kyo-grpc` integration module with ScalaPB-based code generation plus minimal unary client/server support and a small example/test?

A few scope questions:

- Is ScalaPB plugin integration the preferred direction, or would you rather avoid adding a codegen plugin surface at first?
- Should the initial deliverable include streaming support, or is unary RPC enough for a first merged step?
- Do you expect this to be a Kyo-native API from the start, or is a thin interop layer acceptable initially?
- Are benchmarks part of the acceptance criteria, given the issue mentions zio-grpc interop performance?

If you can confirm the smallest acceptable slice, I can break the work down and avoid sending an oversized PR.

**Comment by fwbrasil** [comment]
folks, this bounty was already reserved for @steinybot. We're getting too much noise due to low-quality AI PRs so I'm closing this issue. Since we haven't heard back from @steinybot for awhile, I'm closing this ticket and related PRs. I had tried to remove the bounty but Algora isn't allowing it for some reason. @calvinlfer can you try removing your bounties?

**Comment by arthurus36-alt** [comment]
Hi! I looked into this issue and came up with a technical approach/solution. Here it is:

Here's a direct technical approach:

1.  **ScalaPB Plugin Development**: Implement a custom `protoc` plugin using `scalapb.compiler.CodeGenerator`. This plugin will read `.proto` service definitions and generate Scala traits for server implementations and corresponding client stubs, tailored to the project's effect system (e.g., `fs2.Stream[F, ?]` or `ZIO.Stream[?, ?, ?]`).
2.  **Minimal Runtime Library**: Develop a slim runtime library that provides:
    *   An HTTP/2 server integration (e.g., leveraging Netty directly or an `http4s` backend) to host the generated gRPC service implementations.
    *   A client-side layer to manage HTTP/2 connections and invoke remote gRPC methods via the generated client stubs.
    *   Focus on efficient, low-overhead stream-based I/O to address potential performance concerns.

**Code Snippet (Conceptual Plugin Entry Point):**

```scala
// In a dedicated sbt subproject for the ScalaPB plugin
package com.myproject.grpc.codegen

import scalapb.compiler.{CodeGenerator, ProtocContext, GeneratedFile}
import com.google.protobuf.compiler.PluginProtos.CodeGeneratorResponse

class MyGrpcGenerator extends CodeGenerator {
  override def name: String = "my_project_grpc_scala"

  override def run(context: ProtocContext): CodeGeneratorResponse = {
    val filesWithServices = context.allFiles.filter(_.is.service.nonEmpty)
    
    val generatedFiles = filesWithServices.flatMap(file =>
      file.service.map { service =>
        // Placeholder for core logic:
        // Generate Scala code for service trait (server-side) and client stub (client-side)
        // using project's preferred effect system (e.g., F[_], Stream[F, _]).
        val serviceCode = s"""
          |package ${file.scalaPackage.getOrElse(context.config.flatPackage)}.services
          |
          |import ${context.config.scalapb.basePackage.getOrElse("com.google.protobuf")}.duration.Duration
          |// ... imports for F[_], Stream[F, _]
          |
          |trait ${service.name}Service[F[_]] {
          |  // Methods based on service.method definitions
          |  // Example: def ${service.method.head.name}(request: ${service.method.head.inputType.name}): F[${service.method.head.outputType.name}]
          |}
          |
          |class ${service.name}Client[F[_]](channel: io.grpc.ManagedChannel) extends ${service.name}Service[F] {
          |  // Client stub implementation using channel and generated ScalaPB messages
          |}
        """.stripMargin
        
        GeneratedFile(
          file.scalaPackage.map(_.replace('.', '/')).getOrElse("") + "/" + service.name + "Grpc.scala",
          serviceCode
        )
      }
    )
    
    CodeGeneratorResponse.newBuilder().addAllFile(generatedFiles.asJava).build()
  }
}
```

---
*If this helped point you in the right direction or saved you time, I am accepting freelance work or open-source tips at ETH: `0x71C7656EC7ab88b098defB751B7401B5f6d8976F`*

**Comment by arthurus36-alt** [comment]
Hi! I looked into this issue and came up with a technical approach/solution. Here it is:

```scala
// build.sbt (or similar build config)
libraryDependencies ++= Seq(
  "com.thesamet.scalapb" %% "scalapb-runtime" % scalapb.compiler.Version.scalapbRuntimeVersion % "protobuf",
  "com.thesamet.scalapb" %% "scalapb-runtime-grpc" % scalapb.compiler.Version.scalapbRuntimeVersion, // Required for gRPC
  "org.typelevel" %% "fs2-grpc-runtime" % "[fs2-grpc-version]", // Or zio-grpc-runtime for ZIO
  "org.typelevel" %% "cats-effect" % "[cats-effect-version]" // For FS2-gRPC effect type
)

// project/plugins.sbt
addSbtPlugin("com.thesamet" % "sbt-protoc" % "[sbt-protoc-version]")
libraryDependencies += "com.thesamet.scalapb" %% "compilerplugin" % scalapb.compiler.Version.scalapbRuntimeVersion

// build.sbt - ScalaPB configuration
Compile / PB.targets := Seq(
  scalapb.gen(grpc = false) -> (Compile / sourceManaged).value,
  scalapb.gen(grpc = true, flatPackage = true)
    .addProtoSpecific(".proto_file_path", "option (scalapb.grpc.fs2_stub_options).effect_type = \"cats.effect.IO\";") // For FS2-gRPC
    .addProtoSpecific(".proto_file_path", "option (scalapb.grpc.fs2_stub_options).stream_type = \"fs2.Stream\";")
    -> (Compile / sourceManaged).value
)

// Example Service Implementation (using FS2-gRPC with Cats Effect)
import cats.effect.IO
import fs2.Stream
import com.example.myproto._ // Generated from .proto

class MyGrpcServiceImpl extends MyServiceFs2Grpc[IO] {
  override def myMethod(request: MyRequest): IO[MyResponse] = IO {
    MyResponse(s"Hello, ${request.name}!")
  }

  override def myStreamingMethod(requests: Stream[IO, MyStreamingRequest]): Stream[IO, MyStreamingResponse] = {
    requests.evalMap(req => IO(MyStreamingResponse(s"Processed: ${req.data}")))
  }
}

// To implement: Use fs2-grpc to build and start the server,
// and create a client stub to make calls.
```

---
*If this helped point you in the right direction or saved you time, I am accepting freelance work or open-source tips at ETH: `0x71C7656EC7ab88b098defB751B7401B5f6d8976F`*
